libtsuba/src/WriteGroup.cpp

```diff
@@ -25,49 +25,74 @@ WriteGroup::Make() {
   return std::unique_ptr<WriteGroup>(new WriteGroup(tag));
 }
 
+bool
+WriteGroup::Drain() {
+  auto op_it = pending_ops_.begin();
+  if (op_it == pending_ops_.end()) {
+    return false;
+  }
+  auto res = op_it->result.get();
+  if (!res) {
+    KATANA_LOG_DEBUG(
+        "async write op for {} returned {}", op_it->location, res.error());
+    errors_++;
+    last_error_ = res.error();
+  }
+  outstanding_size_ -= op_it->accounted_size;
+  pending_ops_.erase(op_it);
+  return true;
+}
+
 Result<void>
 WriteGroup::Finish() {
-  Result<void> return_val = katana::ResultSuccess();
-  uint32_t errors = 0;
-
-  for (AsyncOp& op : pending_ops_) {
-    auto res = op.result.get();
-    if (!res) {
-      KATANA_LOG_DEBUG(
-          "async write op for {} returned {}", op.location, res.error());
-      errors++;
-      return_val = res.error();
-    }
+  while (Drain()) {
```

Comment:
perhaps busy waiting is fine for now, but perhaps a condition variable or explicit bounded queue (with waiting) might be good later.